code/modules/wiremod/port.dm

```diff
@@ -0,0 +1,281 @@
+		return QDEL_HINT_LETMELIVE
+	connected_component = null
+	return ..()
+
+/**
+ * Returns the value to be set for the port
+ *
+ * Used for implicit conversions between outputs and inputs (e.g. number -> string)
+ * and applying/removing signals on inputs
+ */
+/datum/port/proc/convert_value(prev_value, value_to_convert)
+	if(prev_value == value_to_convert)
+		return prev_value
+	. = value_to_convert
+
+	switch(datatype)
+		if(PORT_TYPE_STRING)
+			// So that they can't easily get the name like this.
+			if(isatom(value_to_convert))
+				return PORT_TYPE_ATOM
+			else
+				return "[value_to_convert]"
+
+	if(isatom(value_to_convert))
+		var/atom/atom_to_check = value_to_convert
+		if(QDELETED(atom_to_check))
+			return null
+
+/**
+ * Sets the datatype of the port.
+ *
+ * Arguments:
+ * * type_to_set - The type this port is set to.
+ */
+/datum/port/proc/set_datatype(type_to_set)
+	datatype = type_to_set
+	color = datatype_to_color()
+	disconnect()
+	if(connected_component)
+		SStgui.update_uis(connected_component)
+
+/**
+ * Disconnects a port from all other ports
+ *
+ * Called by [/obj/item/circuit_component] whenever it is disconnected from
+ * an integrated circuit
+ */
+/datum/port/proc/disconnect()
+	SHOULD_CALL_PARENT(TRUE)
+	SEND_SIGNAL(src, COMSIG_PORT_DISCONNECT)
+
+
+/**
+ * # Output Port
+ *
+ * An output port that many input ports can connect to
+ *
+ * Sends a signal whenever the output value is changed
+ */
+/datum/port/output
+	/// The output value of the port
+	var/output_value
+
+/datum/port/output/disconnect()
+	set_output(null)
+	return ..()
+
+/datum/port/output/Destroy(force)
+	output_value = null
+	return ..()
+
+/**
+ * Sets the output value of the port
+ *
+ * Arguments:
+ * * value - The value to set it to
+ */
+/datum/port/output/proc/set_output(value)
+	if(isatom(output_value))
+		UnregisterSignal(output_value, COMSIG_PARENT_QDELETING)
+	output_value = convert_value(output_value, value)
+	if(isatom(output_value))
+		RegisterSignal(output_value, COMSIG_PARENT_QDELETING, .proc/null_output)
+
+	SEND_SIGNAL(src, COMSIG_PORT_SET_OUTPUT, output_value)
+
+/// Signal handler proc to null the output if an atom is deleted. An update is not sent because this was not set.
+/datum/port/output/proc/null_output(datum/source)
+	SIGNAL_HANDLER
+	if(output_value == source)
+		output_value = null
+
+/datum/port/output/set_datatype(type_to_set)
+	. = ..()
+	set_output(null)
+
+/**
+ * Determines if a datatype is compatible with this port.
+ *
+ * Arguments:
+ * * other_datatype - The datatype to check
+ */
+/datum/port/output/proc/compatible_datatype(datatype_to_check)
+	if(datatype_to_check == datatype)
+		return TRUE
+
+	if(datatype == PORT_TYPE_NUMBER)
+		// Can easily convert a number to string. Everything else has to use a tostring component
+		return datatype_to_check == PORT_TYPE_STRING
+	if(datatype == PORT_TYPE_SIGNAL)
+		// A signal port is just a number port but distinguishable
+		return datatype_to_check == PORT_TYPE_NUMBER
```

Comment:
Kind of odd thing here, this allows SIGNAL outputs to plug into NUMBER inputs but not visa versa, so you can't for example use a clock NUMBER output to trigger the Compare SIGNAL input on the Comparison circuit.

Reply:
my fault, forgot to change the clock number output to a signal one